Client/Frontend/Browser/BrowserViewController/OpenSearch/BrowserViewController+OpenSearch.swift

```diff
@@ -0,0 +1,212 @@
+        // thus in case of yahoo.com the title is 'Yahoo Search' and Shortname is 'Yahoo'
+        // Instead we are checking referenceURL match to determine searchEngine is added or not
+        
+        let matches = self.profile.searchEngines.orderedEngines.filter {$0.referenceURL == referenceObject.reference}
+        
+        if !matches.isEmpty {
+            self.customSearchEngineButton.state = .disabled
+        } else {
+            self.customSearchEngineButton.state = .enabled
+        }
+        
+        /*
+         This is how we access hidden views in the WKContentView
+         Using the public headers we can find the keyboard accessoryView which is not usually available.
+         Specific values here are from the WKContentView headers.
+         https://github.com/JaviSoto/iOS9-Runtime-Headers/blob/master/Frameworks/WebKit.framework/WKContentView.h
+         */
+        guard let webContentView = UIView.findSubViewWithFirstResponder(webView) else {
+            /*
+             In some cases the URL bar can trigger the keyboard notification. In that case the webview isnt the first responder
+             and a search button should not be added.
+             */
+            return
+        }
+        
+        guard let input = webContentView.perform(#selector(getter: UIResponder.inputAccessoryView)),
+              let inputView = input.takeUnretainedValue() as? UIInputView,
+              let nextButton = inputView.value(forKey: "_nextItem") as? UIBarButtonItem,
+              let nextButtonView = nextButton.value(forKey: "view") as? UIView else {
+            return
+        }
+        
+        inputView.addSubview(customSearchEngineButton)
+        
+        customSearchEngineButton.snp.remakeConstraints { make in
+            make.leading.equalTo(nextButtonView.snp.trailing).offset(20)
+            make.width.equalTo(inputView.snp.height)
+            make.top.equalTo(nextButtonView.snp.top)
+            make.height.equalTo(inputView.snp.height)
+        }
+    }
+
+    @objc func addCustomSearchEngineForFocusedElement() {
+        guard var referenceURLString = openSearchEngine?.reference,
+              let title = openSearchEngine?.title,
+              var referenceURL = URL(string: referenceURLString),
+              let faviconURLString = self.tabManager.selectedTab?.displayFavicon?.url,
+              let iconURL = URL(string: faviconURLString) else {
+            let alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+            present(alert, animated: true, completion: nil)
+            return
+        }
+                
+        guard let scheme = tabManager.selectedTab?.webView?.url?.scheme,
+              let host = tabManager.selectedTab?.webView?.url?.host else {
+            log.error("Selected Tab doesn't have URL")
+            return
+        }
+        
+        while referenceURLString.hasPrefix("/") {
```

Comment:
Why do this? If you're going to reconstruct it from `scheme://host/path` you can probably get away with using `referenceURL.path` and reconstruct it or use `URLComponent`?

Reply:
Honestly the problem is there is no convention about what we might receive as reference.
Sometimes you need to construct and sometimes you can fetch string starts // or / and different types.

Here are some of the examples:
```
href: /w/opensearch_desc.php
href: https://cdn.ecosia.org/opensearch.xml
href: lilosearch.xml
href: /opensearch.xml
href: //opensearch_desc.xml
```